src/Libraries/Microsoft.Extensions.AI/ChatCompletion/ChatClientBuilder.cs

```diff
@@ -41,27 +46,61 @@ public ChatClientBuilder(Func<IServiceProvider, IChatClient> innerClientFactory)
     /// <returns>An instance of <see cref="IChatClient"/> that represents the entire pipeline.</returns>
     public IChatClient Build(IServiceProvider? services = null)
     {
-        services ??= EmptyServiceProvider.Instance;
-        var chatClient = _innerClientFactory(services);
+        ClientFactory chatClientFactory = (f, s) => _innerClientFactory(s);
 
         // To match intuitive expectations, apply the factories in reverse order, so that the first factory added is the outermost.
-        if (_clientFactories is not null)
+        if (_clientMiddleware is not null)
         {
-            for (var i = _clientFactories.Count - 1; i >= 0; i--)
+            for (var i = _clientMiddleware.Count - 1; i >= 0; i--)
             {
-                chatClient = _clientFactories[i](chatClient, services);
-                if (chatClient is null)
+                chatClientFactory = _clientMiddleware[i](chatClientFactory);
+
+                if (chatClientFactory is null)
                 {
                     Throw.InvalidOperationException(
-                        $"The {nameof(ChatClientBuilder)} entry at index {i} returned null. " +
-                        $"Ensure that the callbacks passed to {nameof(Use)} return non-null {nameof(IChatClient)} instances.");
+                        $"The {nameof(ChatClientBuilder)} middleware entry at index {i} returned null. " +
+                        $"Ensure that the callbacks passed to {nameof(Use)} return non-null {nameof(Func<IFeatureCollection, IChatClient>)} instances.");
                 }
             }
         }
 
+        var features = new FeatureCollection();
```

Comment:
What would this be without M.E.Features... `Dictionary<Type, object?>`?

Reply:
I think so. I might make the object non-nullable, but you're always going to have to cast anyway.